# 필요한 라이브러리 불러오기

필요한 라이브러리를 불러오기 위해 다음의 코드를 실행하세요.

In [ ]:
# 그림그리기 위한 matplotlib 라이브러리 불러오기
import matplotlib.pyplot as plt
from matplotlib import rc
import seaborn as sns

In [ ]:
# 한글깨지지 않게 하기 위한 코드
import sys

# Google Colab 환경에서 실행 중인지 확인
if 'google.colab' in sys.modules:
    # debconf를 Noninteractive 모드로 설정
    !echo 'debconf debconf/frontend select Noninteractive' | \
    debconf-set-selections

    # fonts-nanum 패키지를 설치
    !sudo apt-get -qq -y install fonts-nanum

    # Matplotlib의 폰트 매니저 가져오기
    import matplotlib.font_manager as fm

    # 나눔 폰트의 시스템 경로 찾기
    font_files = fm.findSystemFonts(fontpaths=['/usr/share/fonts/truetype/nanum'])

    # 찾은 각 나눔 폰트를 Matplotlib 폰트 매니저에 추가
    for fpath in font_files:
        fm.fontManager.addfont(fpath)

Selecting previously unselected package fonts-nanum.
(Reading database ... 126102 files and directories currently installed.)
Preparing to unpack .../fonts-nanum_20200506-1_all.deb ...
Unpacking fonts-nanum (20200506-1) ...
Setting up fonts-nanum (20200506-1) ...
Processing triggers for fontconfig (2.13.1-4.2ubuntu5) ...


In [ ]:
plt.rcParams['font.family'] = 'NanumGothic'    #사용 방법1
plt.rc('font', family='NanumBarunGothic', size=11) #사용 방법2
print(plt.rcParams['font.family'], plt.rcParams['font.size'])   # 폰트확인

['NanumBarunGothic'] 11.0


In [ ]:
# 데이터를 저장, 처리하기위한 pandas 라이브러리 불러오기
import pandas as pd
# 숫자 연산을 효율적으로 하기 위한 라이브러리 불러오기
import numpy as np

In [ ]:
# 선형회귀모형을 위한 라이브러리
import statsmodels
import statsmodels.api as sm
import statsmodels.formula.api as smf
import statsmodels.stats.api as sms

# 오전 실습 문제

실습문제의 데이터셋을 불러오기 위해 다음 코드를 실행하세요.

In [ ]:
pip install ucimlrepo

In [ ]:
# https://archive.ics.uci.edu/dataset/9/auto+mpg 에서 데이터셋 다운로드

# 데이터 URL (UCI Repository)
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/auto-mpg/auto-mpg.data"

columns = ["mpg", "cylinders", "displacement", "horsepower", "weight",
           "acceleration", "model_year", "origin", "car_name"]

mileage = pd.read_csv(url, delim_whitespace=True, names=columns, na_values='?')
mileage.dropna(inplace=True)
mileage["horsepower"] = mileage["horsepower"].astype(float)

<ipython-input-11-04574066ade4>:9: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\s+'`` instead
  mileage = pd.read_csv(url, delim_whitespace=True, names=columns, na_values='?')


In [ ]:
# 불러와진 데이터셋은 다음과 같습니다.
mileage

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0          8         307.0       130.0  3504.0          12.0   
1    15.0          8         350.0       165.0  3693.0          11.5   
2    18.0          8         318.0       150.0  3436.0          11.0   
3    16.0          8         304.0       150.0  3433.0          12.0   
4    17.0          8         302.0       140.0  3449.0          10.5   
..    ...        ...           ...         ...     ...           ...   
393  27.0          4         140.0        86.0  2790.0          15.6   
394  44.0          4          97.0        52.0  2130.0          24.6   
395  32.0          4         135.0        84.0  2295.0          11.6   
396  28.0          4         120.0        79.0  2625.0          18.6   
397  31.0          4         119.0        82.0  2720.0          19.4   

     model_year  origin                   car_name  
0            70       1  chevrolet chevelle malibu  
1            70       1          buick skylark 320  
2            70       1         plymouth satellite  
3            70       1              amc rebel sst  
4            70       1                ford torino  
..          ...     ...                        ...  
393          82       1            ford mustang gl  
394          82       2                  vw pickup  
395          82       1              dodge rampage  
396          82       1                ford ranger  
397          82       1                 chevy s-10  

[392 rows x 9 columns]

위의 mileage 데이터셋은 자동차의 연비(MPG), 마력(horsepower), 무게(weight) 등 여러 변수를 포함하고 있는 데이터셋입니다. 변수 설명은 다음과 같습니다:

- weight: 자동차의 무게
- mpg: 자동차의 연비

자동차의 무게와 연비가 어떠한 함수를 따르는지를 추정하고자 합니다.

### 문제 A. 자동차의 무게의 일차식을 통해 연비를 예측하는 회귀모형을 만들고, 그 회귀모형의 회귀계수를 구하세요.

Intercept    46.216525
weight       -0.007647
dtype: float64

### 문제 B. 자동차의 무게를 x축으로 하고 연비를 y축으로 하는 산점도를 그리고 문제 A에서 구한 회귀모형을 그 위에 겹쳐그린 뒤, 구한 회귀모형이 두 변수의 관계를 설명하기에 적절한지 판단하세요.

### 문제 C. 문제 A-B에서 구한 회귀모형의 결정계수를 구하세요.

### 문제 D. 자동차의 무게와 연비 사이의 관계를 설명하는 더 적절한 함수관계가 존재한다고 생각한다면 자유롭게 만들어본 뒤 그 모델의 산점도와 회귀선을 그리고 결정계수를 구하세요.

### 문제 E. 문제 A-B에서 만든 모형과 문제 D에서 만든 모형 중 어느 것이 두 변수 사이의 관계를 더 잘 설명한다고 생각하는지를 산점도와 결정계수 두가지를 모두 근거로써 사용하여 판단하세요.

# 오후 실습 문제

실습문제의 데이터셋을 불러오기 위해 다음의 코드를 실행하세요.

In [ ]:
pip install ucimlrepo

In [ ]:
# https://archive.ics.uci.edu/dataset/45/heart+disease 에서 데이터셋 다운로드
from ucimlrepo import fetch_ucirepo

# fetch dataset
heart_disease = fetch_ucirepo(id=45)

# data (as pandas dataframes)
X = heart_disease.data.features
y = heart_disease.data.targets
y[np.isin(y, [1, 2, 3, 4])] = 1

heart = pd.concat([X, y], axis=1)
heart = heart.rename(columns={'num': 'y'})
heart = heart[['sex', 'trestbps', 'chol', 'fbs', 'thalach', 'y']]

<ipython-input-9-34fd6e98ee05>:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  y[np.isin(y, [1, 2, 3, 4])] = 1
<ipython-input-9-34fd6e98ee05>:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  y[np.isin(y, [1, 2, 3, 4])] = 1


In [ ]:
# 불러와진 데이터셋은 다음과 같습니다.
heart

sex  trestbps  chol  fbs  thalach  y
0      1       145   233    1      150  0
1      1       160   286    0      108  1
2      1       120   229    0      129  1
3      1       130   250    0      187  0
4      0       130   204    0      172  0
..   ...       ...   ...  ...      ... ..
298    1       110   264    0      132  1
299    1       144   193    1      141  1
300    1       130   131    0      115  1
301    0       130   236    0      174  1
302    1       138   175    0      173  0

[303 rows x 6 columns]

서울대 병원에서는 심장병 의심 증상이 있어서 내원한 환자들의 심장병 유무를 빠르게 진단하기 위한 모델을 만들고자 합니다. 이를 위해 다음의 6가지 변수들이 얻어졌습니다.

- 1. sex: 성별(여성: 0, 남성: 1), 범주형 변수.
- 2. trestbps: 안정 시의 혈압, 수치형 변수.
- 3. chol: 혈청 콜레스테롤, 수치형 변수.
- 4. fbs: 공복혈당이 120 mg/dl보다 큰지의 여부(크면 1, 작으면 0), 범주형 변수.
- 5. thalach: 최대 심박수, 수치형 변수.
- 6. y: 심장병 유무(심장병이 있으면 1, 없으면 0), 범주형 변수.

1-5번의 변수들을 사용해서 6번 변수인 심장병 유무를 진단하고자 합니다. 이를 위해 303명의 데이터를 수집하였다고 가정합니다. 이에 대한 데이터가 heart라는 이름의 데이터프레임으로 저장되어있습니다.

### 문제 A. 성별 차이가 심장병 유무와 연관이 있는지를 시각적으로 확인하려고 합니다. 성별별 심장병을 가진 사람 수를 히스토그램을 통해 시각화하고 성별별로 어떠한 차이가 있는지를 판단하세요.

### 문제 B. 주어진 heart 데이터셋을 훈련용 데이터와 검증용 데이터로 나눈 뒤 훈련용 데이터에 로지스틱 회귀모델을 적용하여 정확도를 계산하세요.

### 문제 C.

(1) 문제 B에서 만든 로지스틱 회귀모델을 가지고 confusion 행렬을 만드세요. \\

(2) 실제 심장병이 있는 사람한테 모델을 적용했을 때 심장병이 있다고 예측한 비율을 구하세요. \\

(3) 실제 심장병이 없는 사람한테 모델을 적용했을 때 심장병이 없다고 예측한 비율을 구하세요.

### 문제 D. 서울대 병원에 심장병 의심 증상을 보이는 환자가 내원했다고 가정합니다. 이 환자의 정보는 다음과 같습니다.

- 성별: 남성
- 안정 시의 혈압: 160
- 혈청 콜레스테롤: 230
- 공복혈당: 125
- 최대 심박수: 150

앞서 만든 모델을 사용하여 이 환자의 심장병 유무를 예측하세요.